### ISSUE TITLE
2.14.7 regression store.query breaks some many2many relationships

### ISSUE THREAD

**Issue by jrjohnson** [emberjs/data#5140]
This issues doesn't exist in 2.13.2 - 2.14.6, it appears for the first time in 2.14.7 and still exists in 2.15.0-beta.4.

Sorry for the meandering ticket and all the leaky domain stuff from using my own API, not really sure what details might be important here. 

I'm doing some pre-loading of relationships in my router to save network requests. In my route I have:
```
afterModel(model) {
  return store.query('objective', {filters: {courses}});
}
```
Which loads all the `objectives` that would be loaded with a `course.get('objectives')`. The goal is for them to then be available in the store without needing to fetch each one individually.

This works as long as each `objective` returned only has a single course in its relationship like
``` 
{
  id: 1
  courses: [1]
}
```

but it seems to fail if there are more than one:
``` 
{
  id: 1
  courses: [1, 2]
}
```

The result ends up being that the loaded `course` model somehow gets confused and thinks it is only linked to some of the objectives. Operations like `{{#each course.objective}}` or `{{course.objectives.length}}` or even `course.hasMany('objectives').ids()` all provide incorrect data.

This *only* happens when the route is visited directly (by refreshing the page or typing in the URL). If a `link-to` is followed (even one that uses an `id` and not an object for the model then the relationships all load. 

I've tried to create a small reproduction of this, unfortunately I couldn't make it work with ember-twiddle but I have breaking code here:

https://github.com/jrjohnson/reproduction-ember-data-query-does-not-update-store
and a running app example at:
https://grader-duck-81038.netlify.com/

I've created three routes:
1. Working - no `afterModel` hook used, works fine
2. Broken - `afterModel` hook used, breaks
3. Working with fewer related items  uses `afterModel` hook, but loads a different `course` object with `objectives` that are only related to itself.

(edited: added info that 2.15.0-beta.4 was tested)
(edited: I'm not sure where I got it wrong originally but this issues actually showed up in 2.14.7 for the first time.)

**Comment by sandstrom** [comment]
Does `afterModel` even fire when you visit from a link? I think it won't, and that's probably why it works in that case.

This isn't an answer to your issue, but if that's the case I'd amend the issue above to remove the section about link-to.

**Comment by jrjohnson** [comment]
Thanks @sandstrom,
afterModel is fired for every request (just verified to be sure), and i'm pretty sure that even model gets fired when you pass a model id and not a model object which is why I structured the `link-to` that way.

I have verified that payloads are identical where applicable, though the core of this issue is that calling `store.query()` breaks something. Since I don't call that in all of the working cases there is no payload to compare.
